src/matrix.hpp

```diff
@@ -0,0 +1,139 @@
+#pragma once
+
+#include "spaces.hpp"
+
+namespace opt {
+
+template <Vector V, std::size_t Cols>
+struct matrix {
+    static constexpr std::size_t rows = std::tuple_size<V>::value;
+    static constexpr std::size_t cols = Cols;
+    using entries_type = scalar_t<V>;
+    using columns_type = V;
+
+    using row_index = typename V::coords_type::size_type;
+    using col_index = std::size_t;
+
+    std::array<V, Cols> columns{};
+
+    // Special member functions defaulted
+    constexpr matrix() = default;
+
+    // Ctor from elements, row major
```

Comment:
Like matrix<blabla>{row_major, {x, y z}}?
I had that and a by column_major tag in my first try, I was unsure how much it was making the syntax heavy. It can be reincluded, ofc